Automattic/woocommerce-payments
# Return upstream HTTP error code from terminal payment endpoint

Fixes #2847

#### Changes proposed in this Pull Request

Replace `502` HTTP status code in the `capture_terminal_payment` endpoint with the one returned by Stripe.

Previously, we were returning a generic 502 HTTP status code for all Stripe API errors. While 502 was not an inappropriate choice, it represented *any* error at gateway or proxy, leading to the loss of the actual status code returned by the upstream API.

For example, when we try to charge more than the amount in payment intent Stripe fails with a 400 HTTP code which signifies that the input(s) received by Stripe were incorrect or in bad format. Ignoring 400 and returning 502
meant that our endpoint's client had to rely on the error message to determine the cause of error.

#### Testing instructions

A scenario is assumed where a user, after placing a COD order, changes their mind and wants an item removed before making payment.

1. Create a new order with two or more non-digital/virtual products. Eg. a Hoodie ( $45 ) and a Polo ( $20 ).
2. When checking out the order, select **Cash on delivery** as the payment method.
3. The order status will hopefully change to **Processing** soon after order creation. If it's On Hold, change to Processing manually.
4. Refund one item from the order, say Hoodie. The total payable amount should now be reduced by $45. ( Note: directly editing an order to remove items is not possible when the status is Processing ).
5. In WooCommerce Mobile App, attempt to collect payment for the above order. It should ultimately fail due to an incorrect payable amount ( #2846 ). Note the HTTP response from `capture_terminal_payment` endpoint in Charles Proxy (or another network traffic analyzer).
6. If testing with WooCommerce mobile app is not possible, make a direct POST request (from a REST client) to `capture_terminal_payment` endpoint with the above order's id and a new payment intent id ( the payment intent must be created* for an amount greater than the payable amount noted in step 4 ). Note the HTTP response in the REST client.
7. Before this change, the HTTP response code was 502. After this change, it should be 400. ( Screenshots below )

\* We can create a manual payment with the option Capture funds later in https://dashboard.stripe.com/test/connect/accounts/acct_YOUR_TEST_ID/payments/new

##### Before

<img width="1391" alt="image" src="https://user-images.githubusercontent.com/1288616/132537323-5a7adac8-2b92-4f3e-92c4-4098a6d2d39a.png">

##### After

<img width="1459" alt="image" src="https://user-images.githubusercontent.com/1288616/132674198-5aa820fb-c818-472a-aa70-0838a180664a.png">

-------------------
